rfd/0135-kube-apps-discovery.md

```diff
@@ -0,0 +1,199 @@
+on the backend, and then will decide on appropriate actions (create/update/delete) that are required on
+the backend to change resources to the desired state.
+The Teleport app service that runs in the same Kubernetes cluster will then react to those changes, and will start/stop proxying
+apps that are selected by appropriate labels.
+
+Name of the created Teleport app will consist of Kubernetes service name, namespace and cluster name, like this:
+`$SERVICE_NAME-$CLUSTER_NAME-$NAMESPACE`. This is done to avoid naming collisions in discovered apps. Though app names
+will be long in that case, there's ongoing work to improve UX when working with long resources names in Teleport, see [RFD 129](https://github.com/gravitational/teleport/pull/27258).
+
+To allow for finer control over exposed service transformation into Teleport app Kubernetes annotations will be used.
+By default we will be exposing services as `tcp` apps, since it's the most general type, but annotation or exposed port's `appProtocol` field
+can be used to specify which protocol to use for the app's URI. 
+
+If Kubernetes service has multiple ports in use, we will expose each port as a separate app, with name of the app including the port's name
+specified in the service. For example, if a service has port 443 exposed with name `tls` and also port `5432` with name `postgres`
+we will create two apps: `testApp-tls-main-cluster-prod`, `testApp-postgres-main-cluster-prod`. Names for ports are mandatory in Kubernetes if more
+than one port is exposed for the service, so we always will be able to use it if we need it. User will be able to use annotation to specify
+preferred port, in that case only app with this port will be created, without port info in the app name.
+
+In order for discovery service to be able to dynamically create/update/delete apps we will add Read/Write permissions 
+for the apps to the built-in system role `Discovery`.
+
+Labels from the Kubernetes service will be copied to the corresponding Teleport app.
+
+## UX
+
+### Discovery
+
+Discovery service will use new `kubernetes` matchers to periodically poll list of services inside the Kubernetes cluster.
+Labels specified in the config will be used to filter out services that should be exposed. Also users will be able to
+specify namespaces to process. Another new field for the configuration of the discovery service will be `kubernetes_cluster` - 
+it specifies kubernetes cluster name in which discovery service is running. It is mandatory if `kubernetes` matchers are
+specified in the config.
+
+```yaml
+## This section configures the Discovery Service
+discovery_service:
+  enabled: yes
+  discovery_group: kube-auto-apps
+  kubernetes_cluster: main-cluster
+  kubernetes:
+    - types: ["app"] # in the future "db" will be possible
+      namespaces: [ "toronto", "porto" ]
+      labels:
+        env: staging
+    - types: ["app"]
+      namespaces: [ "seattle", "oakland" ]
+      labels:
+        env: testing
```

Comment:
If we go the auto-magical way and require no annotation, it might be useful to let users set a label selector (which defaults to *:*) to control which services they want to pickup.

Reply:
Yep, this is what is shown here though, `labels` allows to filter out services to pick up.

Reply:
my bad, I thought the labels were only meant to be set on the discovered application

Reply:
It will be both :)
First we use these labels to pick up services and then since kube labels from service are copied to the app they will appear on the discovered application as well.